pkg/apis/core/validation/validation.go

```diff
@@ -3165,6 +3167,20 @@ func validateContainerCommon(ctr *core.Container, volumes map[string]core.Volume
 		allErrs = append(allErrs, field.Required(path.Child("image"), ""))
 	}
 
+	// Adding API validation for images using the dockerref vendored package
+	// Main cases to detect are where the repository name is not lowercase or repository name includes 64-byte hexadecimal strings
+	// Other cases may fail the docker check but not sure if all are invalid cases.
+	// A lot of unit tests also assume that image name does not have to be specified
+	if len(ctr.Image) > 0 {
+		_, err := dockerref.ParseNormalizedNamed(ctr.Image)
+		if err != nil {
+			if strings.Contains(err.Error(), "lowercase") {
+				allErrs = append(allErrs, field.Invalid(path.Child("image"), ctr.Image, "repository name must be all lowercase"))
+			} else if strings.Contains(err.Error(), "hexadecimal strings") {
+				allErrs = append(allErrs, field.Invalid(path.Child("image"), ctr.Image, "repository name must not specify 64-byte hexadecimal strings"))
+			}
```

Comment:
Do we ignore other errors in purpose? If so, why?

Reply:
Yea some of just so I can make sure all unit tests pass.  

@derekwaynecarr had a comment in this code about padded whitespace being a case that we shouldn't check: 	
```
// TODO: do not validate leading and trailing whitespace to preserve backward compatibility.
	// for example: https://github.com/openshift/origin/issues/14659 image = " " is special token in pod template
	// others may have done similar
```
I found that our unit tests allow empty image names so I don't check for that and I also avoid this check if we have padded whitespace.  

It is interesting that we actually check and fail this for Ephermeral Containers but not for normal containers.  

I pushed a change to spell out more cases and report user friendly errors:

The following cases are test cases I added:
1) capital letters in the image name.
2) sha256 in the image name
3) http or https in the image name
4) and general error that fails the docker check

Reply:
I'm not sure it makes sense to double check errors returned by `ParseNormalizedNamed` in geenral.
What would happen if `ParseNormalizedNamed` error messages are changed or new checks are added?

Reply:
I'm not sure of the best approach here.  

https://github.com/kubernetes/kubernetes/pull/115736#discussion_r1104833989 was one suggestion to better align these validation messages with the current approach.  

In general I think we would want more descriptive error messages if this validation fails.  The `ParseNormalizedNamed` will return `invalid reference format` for all these cases.  We double check errors so we could give a more descriptive error in the validation message.  

From a user I usually find descriptive errors to be much more helpful than a general one.  

I do have unit tests around this function so that we can make sure that we translate cases to error messages

Reply:
> ParseNormalizedNamed will return invalid reference format for all these cases.

It will return "invalid reference format: repository name must be lowercase" or "invalid repository name (%s), cannot specify 64-byte hexadecimal strings" or "reference %s has no name", which are quite descriptive from my point of view. If you find them less descriptive than needed I'd suggest to submit a PR for `github.com/docker/distribution/digestset`.

If new checks are added to the`ParseNormalizedNamed` API we wouldn't notice it here and consider them as "image is an invalid reference type". This is not we want, I believe.

> https://github.com/kubernetes/kubernetes/pull/115736#discussion_r1104833989 was one suggestion to better align these validation messages with the current approach

Would it help to include error message returned by the `ParseNormalizedNamed` into the Kubelet error message?

Reply:
I pushed a change where I just return the error (if we got one) into the validation message.  You are correct.  I think that is descriptive enough.  And its much cleaner.

Reply:
Thanks. Are there other places in the k/k codebase where image validation is performed? If so, can those validation checks be somehow unified or even dropped?

Reply:
@thockin and you have very similar thoughts!

https://github.com/kubernetes/kubernetes/pull/115736#issuecomment-1428449792

I'm not as sure about this code.  Is it appropriate to apply a default tag in the validation logic?  

I could probably move this utility to validation.go and maybe remove it from kubelet image_manager but not sure what problems that would cause?

Reply:
I was actually thinking about this code: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_image.go#L35
Kubeadm may also have the code that pulls images. Other k/k components can also have something similar.

Reply:
I'm still pretty new to Kubernetes so I would need help on where to look.  Where is the code for Kubeadm?  

I found out that `ParseImageName` is used in most places and we only use a special case in image_manager because we don't want the docker url.  So the author uses `ParseNormalizedNames` so they can ignore the fact that `ParseImageName` returns the full docker repo name.  

I updated the PR.

Reply:
This raises an interesting point!

Kubelet can create Pods from static files on disk, so it needs to do its own validation, but I don't honestly know if it is better to create such a pod with an error condition (which is very visible to the user) or to fail the creation, which will be buried in the kubelet logs.

Perhaps this is why `InvalidImageName` is implemented as it is today?  I don't know the history of that, and we should be careful not to break a UX that was, maybe, intentional.

@derekwaynecarr or @dchen1107 or @smarterclayton may recall the history?

Otherwise we should do some archaeology (git blame + git log) to see if there were discussions about this.

I hope I didn't send you on a wild goose chase, but non-zero chance...

Reply:
Maybe I should put this PR up as a topic of discussion in sig-node.